scheduler/src/cook/mesos/rebalancer.clj

```diff
@@ -391,64 +391,66 @@
     (loop [state init-state
            remaining-preemption max-preemption
            [pending-job-ent & jobs-to-make-room-for] jobs-to-make-room-for
-           pending-job-ents-to-run []
-           task-ents-to-preempt []]
+           preemption-decisions []]
       (if (and pending-job-ent (pos? remaining-preemption))
         (let [[state' preemption-decision] (compute-next-state-and-preemption-decision offer-cache state params pending-job-ent)]
           (if preemption-decision
             (recur state'
                    (dec remaining-preemption)
                    jobs-to-make-room-for
-                   (conj pending-job-ents-to-run pending-job-ent)
-                   (into task-ents-to-preempt
-                         ;; If a task has no id, it must be a synthetic task.
-                         ;; This means that on one iteration of (compute-next-state-and-preemption-decision),
-                         ;; the rebalancer decided to make room for a certain hypothetical task,
-                         ;; but on a subsequent iteration, it became clear that OTHER hypothetical tasks would be an even better outcome.
-                         ;; We shouldn't try to actually preempt tasks that were never scheduled.
-                         (filter :db/id (:task preemption-decision))))
+                   (conj preemption-decisions (assoc preemption-decision
+                                                     :to-make-room-for pending-job-ent)))
             (recur state'
                    remaining-preemption
                    jobs-to-make-room-for
-                   pending-job-ents-to-run
-                   task-ents-to-preempt)))
-
+                   preemption-decisions)))
         (do
           (timers/stop timer)
-          (histograms/update! task-counts-to-preempt (count task-ents-to-preempt))
-          (histograms/update! job-counts-to-run (count pending-job-ents-to-run))
-          ;; pending-job-ents-to-run is only for debugging purpose
-          [pending-job-ents-to-run task-ents-to-preempt]
-          )))))
+          (histograms/update! task-counts-to-preempt (count (mapcat :task preemption-decisions)))
```

Comment:
How much faster? 

I think the current version is more readable and since this is only done once per rebalancer I'm hesitant to change it unless it is too slow.